client/src/main/java/org/eclipse/hono/client/impl/HonoConnectionImpl.java

```diff
@@ -633,10 +633,20 @@ public void closeAndFree(
                     log.debug("sender open [target: {}, sendQueueFull: {}]", targetAddress, sender.sendQueueFull());
                     // wait on credits a little time, if not already given
                     if (sender.getCredit() <= 0) {
-                        vertx.setTimer(clientConfigProperties.getFlowLatency(), timerID -> {
-                            log.debug("sender [target: {}] has {} credits after grace period of {}ms", targetAddress,
-                                    sender.getCredit(), clientConfigProperties.getFlowLatency());
-                            result.tryComplete(sender);
+                        final long waitOnCreditsTimerId = vertx.setTimer(clientConfigProperties.getFlowLatency(),
+                                timerID -> {
+                                    log.debug("sender [target: {}] has {} credits after grace period of {}ms",
+                                            targetAddress,
+                                            sender.getCredit(), clientConfigProperties.getFlowLatency());
+                                    sender.sendQueueDrainHandler(null);
+                                    result.tryComplete(sender);
+                                });
+                        sender.sendQueueDrainHandler(replenishedSender -> {
+                            log.debug("sender [target: {}] has received {} initial credits",
+                                    targetAddress, replenishedSender.getCredit());
+                            vertx.cancelTimer(waitOnCreditsTimerId);
+                            replenishedSender.sendQueueDrainHandler(null);
```

Comment:
this is not necessary as the registered handler will be invoked *exactly once* only anyway.

Reply:
Where do you have this from? Can't find this in proton java docs or from quick view into the implementation. In https://github.com/eclipse/hono/blob/cc645489961ace208449cbbcb0788267ff5897c6/client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java#L263-L278 this behaviour seems to be implemented on top?

Reply:
https://github.com/eclipse/hono/blob/1ef2c6af172b66771b4c8d712a14907a01086622/client/src/main/java/org/eclipse/hono/client/CreditBasedSender.java#L32

Reply:
But I set it on a `ProtonSender`.

Reply:
Good catch :+1: 
My bad. Then you should definitely clear the drain handler like you already do!